src/test/mds/TestQuiesceDb.cc

```diff
@@ -0,0 +1,824 @@
+            if (epoch == this->epoch) {
+              if (this->managers.contains(recipient)) {
+                dout(10) << "listing from the leader (" << leader << ") to " << recipient << " for version " << listing.db_version << " with " << listing.sets.size() << " sets" << dendl;
+                this->managers[recipient]->submit_listing(listing);
+                comms_cond.notify_all();
+                return 0;
+              }
+            }
+            return -1;
+          },
+          [epoch = this->epoch, this, leader, me = this_rank](auto diff_map) {
+            std::unique_lock l(comms_mutex);
+            if (epoch == this->epoch) {
+              if (this->managers.contains(leader)) {
+                dout(10) << "ack from " << me << " to the leader (" << leader << ") for version " << diff_map.version << " with " << diff_map.roots.size() << " roots" << dendl;
+                auto [it, inserted] = latest_acks.insert({me, diff_map});
+                if (!inserted) {
+                  if (it->second.version == diff_map.version) {
+                    if (it->second.roots == diff_map.roots) {
+                      dout(1) << "WARNING: detected a potentialy redundant ack" << dendl;
+                    }
+                  }
+                  it->second = diff_map;
+                }
+                this->managers[leader]->submit_ack_from(me, diff_map);
+                comms_cond.notify_all();
+                return 0;
+              }
+            }
+            return -1;
+          }
+        };
+        mgr->update_membership(mem);
+      }
+      dout(6) << "\n === configured cluster with the following members, starting with the leader: " << leader_and_replicas << dendl;
+    }
+
+    struct TestRequestContext: public QuiesceDbManager::RequestContext, public C_SaferCond {
+      void finish(int r) override { C_SaferCond::finish(r); }
+      void complete(int r) override { C_SaferCond::complete(r); }
+
+      const QuiesceDbTest& parent;
+      TestRequestContext(const QuiesceDbTest& parent) : parent(parent) {}
+      ~TestRequestContext() {
+        wait();
+      }
+
+      bool start(const QuiesceDbRequest::Configurator& c)
+      {
+        done = false;
+        response.clear();
+        request.reset(c);
+
+        int rr = -1;
+
+        for (auto& [rank, mgr] : parent.managers) {
+          if (!(rr = mgr->submit_request(this))) {
+            break;
+          }
+        }
+
+        if (rr == EPERM) {
+          // change the error to something never returned for a request
+          // EPIPE seems reasonable as we couldn't find the leader to send the command to
+          complete(EPIPE);
+          return false;
+        }
+
+        return true;
+      }
+
+      GenericVerboseErrorCode check_result() {
+        std::unique_lock l{lock};
+        if (done) {
+          return ERR(rval);
+        }
+        // this error is never returned by the manager
+        return NA();
```

Comment:
Where is NA() defined? Is this from gtest? I'm not familiar with it.

Reply:
It's defined in the fixture class, for conveniency

Reply:
```C++
    static GenericVerboseErrorCode ERR(int val) {
      return GenericVerboseErrorCode(val);
    }
    static GenericVerboseErrorCode OK()
    {
      return ERR(0);
    }
    static GenericVerboseErrorCode NA() {
      return ERR(EBUSY);
    }
```